### DISCUSSION TITLE
General service relation lost when load balance server involved in trace context

### DISCUSSION THREAD

**Discussion by IceSoda177** [apache/skywalking#10567] [General]
Nginx is playing role as load balance server in our micro service env. After we put the lua agent in nginx, the call relation between  services are inconsistent. For example, the call-in traffic on server side is aggreagted as the server is called by only nginx. And the situation can be worse, if one nginx proxy traffic for different services or the call-in traffic pass two or more nginxs before it reach the server.  
My suggestion is 
1, Add load balance layer, and mark nginx as service of this layer
2, LB layer keep parent exit span info from GENERAL layer in SegmentReference and pass it to next service
3, GENERAL layer entry span can recieve and report two parent span one is from LB layer, the other is from GENERAL layer.
4, OAP could calculate two type of service relation, relation between GENERAL layer and relation between GENERAL layer and LB layer.
5, RPC server return with serviceId, instanceId and endpoint, so the client side of RPC call could know which service or endpoint is handling it's request.

**Comment by wu-sheng** [comment]
> After we put the lua agent in nginx, the call relation between services are inconsistent. For example, the call-in traffic on server side is aggreagted as the server is called by only nginx

Are you calling this an issue? This is what is happening. I don't think your suggestions make any sense.

**Comment by IceSoda177** [comment]
For now, it's just issue for me. NG seperate RPC client and server, and the call relation between general services become uncertain.

**Reply by wu-sheng** [reply]
That is what the proxy brings to you. And it should be like this in a distributed system. SkyWalking as an APM is showing what the system is doing, so, there is nothing we can do, even nothing we should do.

**Reply by IceSoda177** [reply]
Agreed, but I think the lost call relation is also important. It shows the real intention of RPC.

**Reply by wu-sheng** [reply]
I don't think so. Nothing lost. What you expected doesn't exist. Why so obsessive about a not existing thing.

**Reply by IceSoda177** [reply]
Just a discussion. Any point of view is valueble for me. Still appreciate.

Here is another thing. I found OAP could generate a lot of endpoint metrics for NG, because NG agent mark url as endpoint, and sometimes parameters can be placed in the path of url. 
For example, url like GET:/song/{album name}/score. OAP will calc endpoint metric for every album.
I think it's more reasonable if ng agent take location as endpoint, or at least support endpoint name config. 

What do you think about this?

**Reply by wu-sheng** [reply]
Maybe, maybe not. It is different from different use cases.
If you want to group endpoints, the backend provides regex and OpenAPI grouping. It costs CPUs, but it could resolve various case requirements.

**Reply by IceSoda177** [reply]
It does cost CPUs and the configuration will be tedious.

**Reply by wu-sheng** [reply]
Anyway, if you want to change source codes, that is open to you. You could change everything you need to be done. The upsteam has to keep in more general way.

**Reply by IceSoda177** [reply]
OpenResty support getting location matching path from ngx.var [https://forum.openresty.us/d/4889-d1f42a1579cf730437e7e9c4a20d4af9/3](url)
Does nginx agent need to support grouping endpoints by locaiton?
